File: src/cti/api/v1/alerts.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from cti.core.database import get_db
from cti.core.dependencies import AnalystUser, CurrentUser
from cti.schemas.alert import (
    AlertEventListResponse,
    AlertEventResponse,
    AlertRuleCreate,
    AlertRuleResponse,
    AlertRuleUpdate,
)
from cti.services import alert_service

router = APIRouter()


# ---------------------------------------------------------------------------
# Alert Rules
# ---------------------------------------------------------------------------


@router.get("/rules", response_model=list[AlertRuleResponse])
async def list_alert_rules(
    _user: CurrentUser,
    db: AsyncSession = Depends(get_db),
) -> list[AlertRuleResponse]:
    """List all alert rules."""
    rules = await alert_service.list_alert_rules(db)
    return [AlertRuleResponse.model_validate(r) for r in rules]


@router.post("/rules", response_model=AlertRuleResponse, status_code=status.HTTP_201_CREATED)
async def create_alert_rule(
    data: AlertRuleCreate,
    user: AnalystUser,
    db: AsyncSession = Depends(get_db),
) -> AlertRuleResponse:
    """Create a new alert rule."""
    rule = await alert_service.create_alert_rule(db, data, user_id=user.id)
    return AlertRuleResponse.model_validate(rule)


@router.put("/rules/{rule_id}", response_model=AlertRuleResponse)
async def update_alert_rule(
    rule_id: uuid.UUID,
    data: AlertRuleUpdate,
    _user: AnalystUser,
    db: AsyncSession = Depends(get_db),
) -> AlertRuleResponse:
    """Update an existing alert rule."""
    rule = await alert_service.update_alert_rule(db, rule_id, data)
    if not rule:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Alert rule not found"
        )
    return AlertRuleResponse.model_validate(rule)


@router.delete("/rules/{rule_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_alert_rule(
    rule_id: uuid.UUID,
    _user: AnalystUser,
    db: AsyncSession = Depends(get_db),
) -> None:
    """Delete an alert rule."""
    deleted = await alert_service.delete_alert_rule(db, rule_id)
    if not deleted:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Alert rule not found"
        )


# ---------------------------------------------------------------------------
# Alert Events
# ---------------------------------------------------------------------------


@router.get("/events", response_model=AlertEventListResponse)
async def list_alert_events(
    _user: CurrentUser,
    db: AsyncSession = Depends(get_db),
    page: int = Query(default=1, ge=1),
    size: int = Query(default=20, ge=1, le=100),
    rule_id: uuid.UUID | None = None,
) -> AlertEventListResponse:
    """List alert events with pagination."""
    items, total = await alert_service.list_alert_events(
        db, page=page, size=size, rule_id=rule_id
    )
    return AlertEventListResponse(
        items=[AlertEventResponse(**item) for item in items],
        total=total,
        page=page,
        size=size,
    )

File: frontend/src/api/alerts.ts
import client from './client';
import type {
  AlertRuleCreate,
  AlertRuleUpdate,
  AlertRuleResponse,
  AlertEventListResponse,
  WebhookConfigCreate,
  WebhookConfigUpdate,
  WebhookConfigResponse,
  WebhookTestResponse,
} from '@/types/alert';

// --- Alert Rules ---

export async function listAlertRules(): Promise<AlertRuleResponse[]> {
  const response = await client.get<AlertRuleResponse[]>('/alerts/rules');
  return response.data;
}

export async function createAlertRule(
  data: AlertRuleCreate
): Promise<AlertRuleResponse> {
  const response = await client.post<AlertRuleResponse>(
    '/alerts/rules',
    data
  );
  return response.data;
}

export async function updateAlertRule(
  id: string,
  data: AlertRuleUpdate
): Promise<AlertRuleResponse> {
  const response = await client.put<AlertRuleResponse>(
    `/alerts/rules/${id}`,
    data
  );
  return response.data;
}

export async function deleteAlertRule(id: string): Promise<void> {
  await client.delete(`/alerts/rules/${id}`);
}

// --- Alert Events ---

export async function listAlertEvents(params: {
  page?: number;
  size?: number;
  rule_id?: string;
}): Promise<AlertEventListResponse> {
  const response = await client.get<AlertEventListResponse>(
    '/alerts/events',
    { params }
  );
  return response.data;
}

// --- Webhooks ---

export async function listWebhooks(): Promise<WebhookConfigResponse[]> {
  const response = await client.get<WebhookConfigResponse[]>('/webhooks');
  return response.data;
}

export async function createWebhook(
  data: WebhookConfigCreate
): Promise<WebhookConfigResponse> {
  const response = await client.post<WebhookConfigResponse>(
    '/webhooks',
    data
  );
  return response.data;
}

export async function updateWebhook(
  id: string,
  data: WebhookConfigUpdate
): Promise<WebhookConfigResponse> {
  const response = await client.put<WebhookConfigResponse>(
    `/webhooks/${id}`,
    data
  );
  return response.data;
}

export async function deleteWebhook(id: string): Promise<void> {
  await client.delete(`/webhooks/${id}`);
}

export async function testWebhook(id: string): Promise<WebhookTestResponse> {
  const response = await client.post<WebhookTestResponse>(
    `/webhooks/${id}/test`
  );
  return response.data;
}
